### DISCUSSION TITLE
Add support for "inclusions" in dependencyManagement

### DISCUSSION THREAD

**Discussion by chrisdutz** [support-and-care/maven-support-and-care#21] [Ideas for Funding]
While dependencyManagment allows exclusion of certain artifacts, it doesn't allow adding replacement dependencies.

For example if a dependency uses log4j and others use reload4j, it would be great if I could manage that the dependencies pulling in log4j would exclude log4j and include reload4j. Right now I could centrally exclude log4j, but I would need to manually add releoad4j to all places where I'm pulling in the dependency. 

Another scenario is that some dependencies pull in some "all-deps" artifacts. I some times need to exclude these and instead import a set of smaller explicit artifacts. Here again I would need to add this set of dependencies to each place pulling in the main dependency.

So my proposal would be something like this:

```
            <dependency>
                <groupId>org.apache.hadoop</groupId>
                <artifactId>hadoop-common</artifactId>
                <version>${hadoop.version}</version>
                <exclusions>
                    <exclusion>
                        <groupId>log4j</groupId>
                        <artifactId>log4j</artifactId>
                    </exclusion>
                </exclusions>
                <inclusions>
                    <inclusion>
                        <groupId>ch.qos.reload4j</groupId>
                        <artifactId>reload4j</artifactId>
                        <version>${reload4j.version}</version>
                    </inclusion>
                </inclusions>
            </dependency>
```

**Comment by chrisdutz** [comment]
Even if this could be considered a "replace", having a separate exclusion and inclusion part could also allow adding optional dependencies to a dependency ... this could then be a dependencyManagement dependency which only has inclusion elements.

For example if there is an application making use of Apache PLC4X, a user could rely on the API module, but have the drivers included in a central dependencyManagement block.

**Comment by hendrikebbers** [comment]
I totally understand your use case, and I have had that problem several times before. As you already mentioned, I want to bring the idea of an "replace" to the table.

As for an `exclusion` you need to add an `inclusion` at several places in your dependency management definition. Especially in a multi-module build that can make the dependency definition very complex to read.

A replacement could be done in the central `dependencyManagement` section of a project only once:

```
<dependencyManagement>
    <dependencies>
        ...
    </dependencies>
    <replacements>
        < replacement>
            <from>
                <dependency>
                    <groupId>org.foo</groupId>
                    <artifactId>bar</artifactId>
                    <version>3.1.5</version>
                </dependency>
            </from>
            <to>
                <dependency>
                    <groupId>org.foo</groupId>
                    <artifactId>bar-java-module-support</artifactId>
                    <version>3.1.5</version>
                </dependency>
            </to>
        < /replacement>
    </replacements>
</dependencyManagement>

```


The tags `from` and `to` can include 0-n `dependency` tags.

**Reply by chrisdutz** [reply]
Ok ... now I understand the difference ... ok ... that also makes sense, even if it scratches a slightly different itch.

As I mentioned in the discussion on slack, one benefit a "inclusion" would have, would be pure-inclusions ... so I could add dependencies to others. But I guess this option could solve most of the problems I was facing in the past.

**Reply by hendrikebbers** [reply]
cool. Should we keep it in this discussion as a possible solution than or should the "replacement" feature another discussion since you want the split that 2 ideas?

**Reply by chrisdutz** [reply]
Guess we can keep it as one. After all ... what I really want is the "replacement" and my proposal of inclusions would be one option.
